В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [10]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

Н0 - середній ріст студентів = 170 см, Н1 - середній ріст студентів > 170 см

In [21]:
from statsmodels.stats.weightstats import ztest
import numpy as np

In [22]:
null_mean = np.mean(heights)
null_mean


np.float64(170.6969696969697)

In [23]:
z_stat, p_value = ztest(heights, value=170, alternative='larger')
z_stat, p_value


(np.float64(0.8482235371587619), np.float64(0.19815674571091746))

Висновок: Середнє вибірки = 170.70 см
Z‑статистика = 0.41, що менше референтного значення 1,6. 
p‑value = 0.34, що менше рівня значущості 0,5, тобто ми не відхиляємо нульову гіпотезу Н0.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [34]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

Розв'язання 2.1

Н0 - середній бал студенетів group_a дорівнює середньому балу group_b.
Н1 - середній бал студенетів group_a вище середньомго балу group_b

In [36]:
group_a_mean = np.mean(group_a_scores)
group_a_mean

np.float64(77.08125)

In [37]:
group_b_mean = np.mean(group_b_scores)
group_b_mean

np.float64(74.73875000000001)

In [38]:
ztest(group_a_scores, group_b_scores, value=0 )

(np.float64(2.357354864751502), np.float64(0.01840564855252801))

Висновок 2.1: Cтатистика для двох вибірок  z-test=2.36, що більше референтного значення  1,6 при коеф.значущості 0.05. p-value  0,02 - менший за рывень начущості 0.05. Таким чином в нас  додастатньо аргументыв відхилити Н0 і прийняти H1 - після оновлення  курсу середній бал групи А є вищим (і середнє, і z-test позитивний і більше 1.6) ніж середній бал групи В, яка навчалвсь по старій програмі. 

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

Розв'язання 2.2

In [43]:
group_a1_scores = [78.55, 72.25, 79.88, 75.  , 76.54]

In [44]:
group_a1_mean = np.mean(group_a1_scores)
group_a1_mean

np.float64(76.444)

In [45]:
from scipy.stats import ttest_ind

t_stat, p_value = ttest_ind(group_a1_scores, group_b_scores, equal_var=False)
t_stat, p_value


(np.float64(1.1226527301193456), np.float64(0.3010736990344312))

Висновок 2.2 При такому наборі даних, навіть за умови, що середенє для цих випадкових 5 значень групи А1 більше, ніж для групи В (76,44 проти 74.74) показники t-статистики (1,12 < 1,6; 0,3 > 0,05) вказують, що ми маємо дотримуваьтсь гіпотези Н0, відхилити Н1. Маленька група не показала значних покращень оновдення курсу. 


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

В двному випадку порівнюємо  не дві вибірки між собою, а одну вибірку різниць:
diff=after−before і перевіряємо, чи середня різниця > 0.

In [48]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [49]:

before_training = np.array([57.82, 37.63, 36.8, 55.22, 52.97, 52.5, 53.46, 43.2, 52.32,
                            52.39, 42.86, 68.66, 54.74, 36.09, 56.47, 50.25, 57.87, 61.59,
                            41.79, 59.63, 53.12, 58.22, 49.63, 53.23, 53.41, 52.77])

after_training = np.array([62.47, 46.08, 42.7, 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
                           55.48, 47.28, 72.6, 57.59, 59.46, 58.42, 62.28, 61.05, 65.29,
                           42.3, 62.98, 57.9, 59.35, 72.28, 56.64, 43.18, 44.82])


len(before_training), len(after_training) # переконуємось, що довжини однакові


(26, 26)

Н0 - різниця оцінок до тренінгу і після тренінгу дорівнює 0
Н1 - різниця оцінок до тренінгу і після тренінгу вище 0

In [50]:
diff = after_training - before_training # вибірка різниць
diff_mean = diff.mean()
diff_mean


np.float64(4.479615384615385)

Робимо одновибірковий Z‑тест для різниць:
H0 середня різниця = 0 (тренінг не вплинув)
H1 середня різниця > 0 (після тренінгу краще)

In [51]:
z_stat, p_value = ztest(diff, value=0, alternative='larger')
z_stat, p_value


(np.float64(3.271021683095811), np.float64(0.0005357984691984685))

Висновок 3. Різниця є статистично значущою, оскільки z-test 3,27 > 1,6. p_value 0,0005 < 0,05. Працівники в середньому покращили результат на 4.5 одиниці